Inventory Rework and clothing

stilnat opened the pull request:
## Summary

This PR introduces some changes in the way the inventory works. 

The main change is that the inventory is dynamic. When wearing a jumpsuit, pocket slots will display in the player UI, when removing it, the slots are no longer available. Same thing happens when a player lose an arm, let's say the left arm, then a Hand slot disappear as well as the left glove slot.

The UI looks the same as before, but it adapts to the containers present on the Player. If the player has 3 hand containers, 3 hand slots will display in the UI.

I tried to make it customizable without too many difficulties, such as the slots position. They appear in any given order with the small constraint that two containers of the same type will have adjacent slots, such as two hands, or two pockets. 

Finally, I added clothing as it was really useful to test this new inventory and it's also visible progress.
For this first iteration, I figured it would be much easier to have simply one slot one item, such as left shoe right shoe, left glove right glove.

## Pictures/Videos

A player missing its left hand, as you can see, it's not displaying the left hand slot nor the left glove slot.
![image](https://github.com/RE-SS3D/SS3D/assets/14344825/a5c5c8fc-9eec-49e5-86e6-5506cd28e312)

Naked player, no pockets ! 
![image](https://github.com/RE-SS3D/SS3D/assets/14344825/317cb362-5bc3-42b0-b4fa-54b808dec97d)

Small showcase video, not the best possible but we see a few things on it already. Pockets don't appear on the grey jumpsuit simply because I didn't put any on it. I forgot and that's why there's some confusion in the video when I put the grey jumpsuit.

https://www.youtube.com/watch?v=Xm1Fw9bgCRc

## Changes to Files

- Add a bunch of filters and traits for clothing 
- Add a bunch of clothing items
- Modify slots prefabs to make them variant prefab of a generic slot to ease adding/modifying slot
- Inventory is renamed HumanInventory. It only handle "slot containers" now, that is container that display as slots in the player UI.
- Add containerViewer, a script to handle logic of opened or closed containers, which are not part of the Player inventory.


## Technical Notes (optional)

<!-- Provide a more technical description of your changes to help save the reviewers some time. -->

## Known issues

- It works on client and on host in editor BUT for some reasons, meshes are broken on build on my side, I've no idea why. It need to be solved before any merge is possible. NEED HELP. SOLVED ! 
- PDA displays weirdly when worn by player. There's a warning about an issue with bones or something regarding its mesh. Solved.
- Wearing a jumpsuit can cause in some conditions the editor to crash, it doesn't seem to happen on build. Maybe solved ?
- When late joining, host PDA on clients appears on the ground, despite being in the host inventory. The client can't interact with it, and when the host moves the PDA around it fixes it, minor issue.

## TODO

- [x] Remove inherited container, replaces with bool IsSlotContainer or something.
- [x] Document every change.
- [x] Fix the mesh bug on build.
- [x] Add a few items (mask, ear stuff, backpack)
- [x] removing a gameobject with a container at run time should remove the slot as well.

stilnat committed:
allow old view inventory to avoid crashing

stilnat committed:
Add type to slots and resize slots

stilnat committed:
make divisor a prefab

stilnat committed:
Add basis for new inventory

stilnat committed:
Add new inventory to human and inventory view to canvas

stilnat committed:
Add alternative inventory container

stilnat committed:
Make alternative dynamic UI work

stilnat committed:
remove handsview and put highlight in hands script

stilnat committed:
require inventoryalt component

stilnat committed:
fix completely container UI interactions

stilnat committed:
fix everything

stilnat committed:
separate inventory into inventory and containerviewer

this alleviates a bit the role of the inventory. Inventory should only handles what's on the player or hold by them. Opened Containers by the player should be displayed using another script, dedicated to that.

stilnat committed:
add doc, add method to InventoryView

stilnat committed:
adds doc

stilnat committed:
rename Inventory to HUmanInventory

this prevents the conflit with Inventory namespace

stilnat committed:
a bit of decoupling and renaming

stilnat committed:
make hands set up after inventory to avoid timing issues

stilnat committed:
fix weird hand highlight bug

stilnat committed:
sync loadout

stilnat committed:
prefab update with some bugs

stilnat committed:
fix slots type and count hand slots

stilnat committed:
kinda fix highlight hands colors

stilnat committed:
remove old hand prefabs

stilnat committed:
add bag slot and fix hand distance

stilnat committed:
fix order of slots

stilnat committed:
add clothing slots and containers

stilnat committed:
add trucker hat and sunglasses

stilnat committed:
add cloth displayers for glasses

stilnat committed:
fix a bit some patch

stilnat committed:
add hat display

stilnat committed:
add some doc

stilnat committed:
factorise some code

stilnat committed:
fix value in container type to have non overlapping bits

stilnat committed:
update human prefab

stilnat committed:
Add some clothes items

stilnat committed:
make pair slots individual, fix container type

stilnat committed:
add display for other type of clothes

stilnat committed:
small fix

stilnat committed:
add inventory slot container

stilnat committed:
remove useless using

stilnat committed:
fix container on jumpsuit

stilnat committed:
make slot remove when container removed

stilnat committed:
add and remove slots and containers in inventory

stilnat committed:
update hack

stilnat committed:
fix exception on item display

stilnat committed:
show item present in slot

stilnat committed:
update human prefab

stilnat committed:
Merge branch 'develop' into rework-inventory-and-clothing

stilnat committed:
sync clothing display

stilnat committed:
remove warning about parent transform

stilnat committed:
allows to remove container in UI of clients

stilnat committed:
slightly reduce the mesh bug

stilnat committed:
Add bunch of traits for clothing

stilnat committed:
add backpack item

stilnat committed:
modify a bit slots

stilnat committed:
remove separator and improve methods to select slots

stilnat committed:
adding id slot and inventory

stilnat committed:
remove shoes

stilnat committed:
rename stuff, document HumanInventory

stilnat committed:
rename, document inventoryview

stilnat committed:
rename, document containerviewer

stilnat committed:
doc an name change in clothesDisplayer

stilnat committed:
small rename

stilnat committed:
Merge branch 'develop' into rework-inventory-and-clothing

stilnat committed:
fix build clothes displaying weirdly

stilnat committed:
add support for bag and face mask

stilnat committed:
add headset item and support

stilnat committed:
remove inventory slot

stilnat committed:
fix hand slots

stilnat committed:
Revert "remove inventory slot"

This reverts commit 07d3b0ec5d576116926c17b0b4b5674e3b7fc742.

stilnat committed:
fix glove left and right

stilnat committed:
fix slot position in horizontal layout

stilnat committed:
remove inventoryslot container

stilnat committed:
only container with display as slot in the UI

stilnat committed:
clicking on wrong containers no longer drop item

stilnat committed:
add some doc and tags

stilnat committed:
Merge branch 'develop' into rework-inventory-and-clothing

joaoburatto commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/ClothesDisplayer.cs:
```diff
@@ -0,0 +1,222 @@
+        [SyncVar(OnChange = nameof(SyncCloth))]
+        private ClothDisplayData _earLeftData;
+
+        [SyncVar(OnChange = nameof(SyncCloth))]
```
Hardcoding these values make little sense for me, can't we have a ClothingSlot enum and use a SyncDictionary to sync these values?

stilnat replied:
probably yes, I can try that if you prefer.

joaoburatto replied:
Please do

joaoburatto commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/HumanInventory.cs:
```diff
@@ -0,0 +1,421 @@
+            }
+        }
+
+        /// <summary>
```
Why is this accessor down here? Move it after field declarations

stilnat committed:
PDa is no longer considered clothing

stilnat committed:
Merge branch 'develop' into rework-inventory-and-clothing

stilnat committed:
fixing hand containers from push and fixing PDA container

stilnat committed:
making containerViewer private

stilnat committed:
change to dynamic list of displayed clothes

stilnat committed:
partly fix some handling positions

stilnat committed:
fix pda spawning multiple times

stilnat committed:
correctly set up viewer only for owner

stilnat committed:
remove useless PDA and methods

stilnat committed:
move accessor

stilnat committed:
remove inventory container on disabled

stilnat committed:
add spawning with clothes

stilnat committed:
fixing forgotten change

stilnat committed:
update assistant loadout

stilnat committed:
add headset to items, fix loadout

stilnat committed:
fix sec jumpsuit

stilnat committed:
fix return too soon

stilnat committed:
fix jumpsuit

stilnat committed:
fix pocket jumpsuit

stilnat committed:
add and remove containers methods

stilnat committed:
removing and adding slots now works properly

When a gameobject container is disabled and was part of the inventory, the container slot linked to it disappear. Similar when adding a container.

stilnat committed:
onenable add container

stilnat committed:
fix slot size of items sprite

stilnat committed:
fixing item missing exceptions

cosmiccoincidence commented:
> 235 files changed.

We should try to not exceed 150 files in a PR. After 100 files, it becomes quite burdensome on the reviewer to have to review so many files in a single PR.

cosmiccoincidence commented:
Gave it a quick playtest and didn't notice anything obvious that I would change.

I would like to see <kbd>E</kbd> to be used for quick equipping of a clothing garment to the appropriate slot, though this should wait for another PR at this point.

stilnat commented:
@cosmiccoincidence I went a bit over the top with the number of files changed, sorry about it, I'm the first advocating against big PRs. For my defense, most of those files are just small created assets or prefabs for items, and the large majority is added stuff, not modified one. On this one I'd like to play the "trust me bro" card, I think around 40 files are worth reviewing (which is already a bit too much imo).

stilnat committed:
Merge branch 'develop' into rework-inventory-and-clothing

stilnat committed:
fix grid

stilnat committed:
put jumpsuit on item layer

stilnat committed:
clean and doc
